vllm/v1/metrics/stats.py

```diff
@@ -10,3 +14,21 @@ class SchedulerStats:
 
     # gpu_cache_usage: float = 0.0
     # gpu_prefix_cache_hit_rate: float = 0.0
+
+
+class IterationStats:
+    """Stats associated with a single set of EngineCoreOutputs."""
+
+    def __init__(self, log_stats: bool):
+        self.log_stats = log_stats
+        self.num_generation_tokens = 0
+        self.num_prompt_tokens = 0
+
+    def update_from_output(self, output: "EngineCoreOutput",
+                           is_prefilling: bool, prompt_len: int):
+        if not self.log_stats:
+            return
+
+        self.num_generation_tokens += len(output.new_token_ids)
+        if is_prefilling:
```

Comment:
I think one scenario that might potentially complicate things is when a request has a pretty long prompt, and its prefill actually will span multiple "iteration". With current architecture, the prompt throughput stats is actually not accurate. 

I wonder if we should have a path where we propogate each scheduler iteration rather than engine core iteration to the front end process for more accurate stats.

Reply:
We don't send outputs from `EngineCore` until it generates a token. So from the POV of the `Engine` `len(new_token_ids) > 0`. We should add an assert for this invariant.

Reply:
added comment and assert.

Reply:
yeah thats true, but we will log these as counters anyways to prometheus so I dont think it would be a problem unless a prefill runs for >15s

Reply:
Whoops, late to the party here but that is a very nice feature of V0 that we log the prompt throughput stats accurately for long prompts. Where products enable really long contexts, we've seen prompts take >1 minute to prefill and everything appears to be hanging except for the logs saying `Avg prompt throughput: 1092.1 tokens/s`. If I understand this correctly, we'd be limited to a minute+ of radio silence while waiting for the prompt to prefill, which would lead to spiky and confusing metrics.

I agree with @rickyyx that there should be a path to accurately tracking prompt throughput